packages/frontend/src/hooks/useKeycard_test.tsx

```diff
@@ -0,0 +1,80 @@
+import "../happyDomSetup.ts";
+import { useEffect } from "react";
+import { useWalletClient } from "wagmi";
+import { assertEquals } from "@std/assert";
+import {
+  cleanup,
+  render,
+  renderHook,
+  screen,
+  waitFor,
+} from "@testing-library/react";
+import { expect } from "@std/expect";
+
+import { random256BigInt } from "@massmarket/utils";
+
+import { createRouterWrapper } from "../testutils/mod.tsx";
+import { useShopId } from "../hooks/useShopId.ts";
+import { useRelayEndpoint } from "../hooks/useRelayEndpoint.ts";
+import { useKeycard } from "./useKeycard.ts";
+import { KeycardRole } from "../types.ts";
+
+const TestComponent = ({ role }: { role: KeycardRole }) => {
+  const { keycard, addKeycard } = useKeycard();
+  const { data: wallet } = useWalletClient();
+  const { shopId } = useShopId();
+  const { relayEndpoint } = useRelayEndpoint();
+
+  useEffect(() => {
+    if (!wallet || !shopId || !relayEndpoint) return;
+    addKeycard(role).then();
+  }, [wallet, shopId, relayEndpoint]);
+
+  return <p data-testid="keycard-role">{keycard?.role}</p>;
+};
+
+Deno.test("useKeycard", {
+  sanitizeResources: false,
+  sanitizeOps: false,
+}, async (t) => {
+  await t.step(
```

Comment:
I think that all the steps here could be top level `Deno.tests` since they all look like independ tests instead of multiple steps in one tests. (i know we have been using `steps` like this  though)